sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/WatchForKafkaTopicPartitions.java

```diff
@@ -0,0 +1,136 @@
+ */
+package org.apache.beam.sdk.io.kafka;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import org.apache.beam.sdk.metrics.Counter;
+import org.apache.beam.sdk.metrics.Metrics;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.Impulse;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.apache.beam.sdk.transforms.Watch;
+import org.apache.beam.sdk.transforms.Watch.Growth.PollFn;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.annotations.VisibleForTesting;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.TopicPartition;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+public class WatchForKafkaTopicPartitions
+    extends PTransform<PBegin, PCollection<KafkaSourceDescriptor>> {
+
+  private static final Duration DEFAULT_CHECK_DURATION = Duration.standardHours(1);
+  private static final String COUNTER_NAMESPACE = "watch_kafka_topic_partition";
+
+  private final Duration checkDuration;
+  private final SerializableFunction<Map<String, Object>, Consumer<byte[], byte[]>>
+      kafkaConsumerFactoryFn;
+  private final Map<String, Object> kafkaConsumerConfig;
+  private final @Nullable SerializableFunction<TopicPartition, Boolean> checkStopReadingFn;
+  private final List<String> topics;
+  private final @Nullable Instant startReadTime;
+  private final @Nullable Instant stopReadTime;
+
+  public WatchForKafkaTopicPartitions(
+      @Nullable Duration checkDuration,
+      SerializableFunction<Map<String, Object>, Consumer<byte[], byte[]>> kafkaConsumerFactoryFn,
+      Map<String, Object> kafkaConsumerConfig,
+      @Nullable SerializableFunction<TopicPartition, Boolean> checkStopReadingFn,
+      List<String> topics,
+      @Nullable Instant startReadTime,
+      @Nullable Instant stopReadTime) {
+    this.checkDuration = checkDuration == null ? DEFAULT_CHECK_DURATION : checkDuration;
+    this.kafkaConsumerFactoryFn = kafkaConsumerFactoryFn;
+    this.kafkaConsumerConfig = kafkaConsumerConfig;
+    this.checkStopReadingFn = checkStopReadingFn;
+    this.topics = topics;
+    this.startReadTime = startReadTime;
+    this.stopReadTime = stopReadTime;
+  }
+
+  @Override
+  public PCollection<KafkaSourceDescriptor> expand(PBegin input) {
+    return input
+        .apply(Impulse.create())
+        .apply(
+            "Match new TopicPartitions",
+            Watch.growthOf(new WatchPartitionFn()).withPollInterval(checkDuration))
+        .apply(ParDo.of(new ConvertToDescriptor()));
+  }
+
+  private class ConvertToDescriptor
```

Comment:
Consider make inner class static per the guideline of DoFn [here](https://github.com/apache/beam/blob/adac54e1fa4f8f5915b47b16fa7b343b29304c07/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ParDo.java#L298). May need to change reference-to-outer class to constructor params.